Input can not restart itself after extractor changes due to port being in use

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behavior
* Have a running input (in my case Beats input)
* Edit GROK extractor on this input
* On extractor change input is stopped and automatically started again

## Current Behavior
* Have a running input (in my case Beats input)
* Edit GROK extractor on this input
* On extractor change input is stopped
* Input can not be restarted due to  "java.net.BindException: Address already in use" error

## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
1. Fresh GL install
2. Add Beats input
3. Ensure input is started
4. Add GROK extractor
5. change grok pattern
6. input is stopped but not restarted

```
Jun 12 14:15:31 hetznix graylogctl[4970]: 2017-06-12 14:15:31,552 INFO : org.graylog2.rest.resources.system.inputs.ExtractorsResource - Updated extractor <cbb424a0-4f78-11e7-a2c4-5e193e146121> of type [grok] in input <593d5937bfc060437b083017>.
Jun 12 14:15:32 hetznix graylogctl[4970]: 2017-06-12 14:15:32,488 INFO : org.graylog2.inputs.InputStateListener - Input [Beats/593d5937bfc060437b083017] is now STOPPING
Jun 12 14:15:32 hetznix graylogctl[4970]: 2017-06-12 14:15:32,513 INFO : org.graylog2.inputs.InputStateListener - Input [Beats/593d5937bfc060437b083017] is now STOPPED
Jun 12 14:15:32 hetznix graylogctl[4970]: 2017-06-12 14:15:32,515 INFO : org.graylog2.inputs.InputStateListener - Input [Beats/593d5937bfc060437b083017] is now TERMINATED
Jun 12 14:15:32 hetznix graylogctl[4970]: 2017-06-12 14:15:32,516 INFO : org.graylog2.inputs.InputStateListener - Input [Beats/593d5937bfc060437b083017] is now STOPPING
Jun 12 14:15:32 hetznix graylogctl[4970]: 2017-06-12 14:15:32,517 INFO : org.graylog2.inputs.InputStateListener - Input [Beats/593d5937bfc060437b083017] is now STARTING
Jun 12 14:15:32 hetznix graylogctl[4970]: 2017-06-12 14:15:32,518 INFO : org.graylog2.inputs.InputStateListener - Input [Beats/593d5937bfc060437b083017] is now STOPPED
Jun 12 14:15:32 hetznix graylogctl[4970]: 2017-06-12 14:15:32,518 INFO : org.graylog2.inputs.InputStateListener - Input [Beats/593d5937bfc060437b083017] is now TERMINATED
Jun 12 14:15:32 hetznix graylogctl[4970]: 2017-06-12 14:15:32,571 WARN : org.graylog2.plugin.inputs.transports.NettyTransport - receiveBufferSize (SO_RCVBUF) for input BeatsInput{title=Global Beats, type=org.graylog.plugins.beats.BeatsInput, nodeId=null} should be 1048576 but is 212992.
Jun 12 14:15:32 hetznix graylogctl[4970]: 2017-06-12 14:15:32,574 INFO : org.graylog2.inputs.InputStateListener - Input [Beats/593d5937bfc060437b083017] is now RUNNING
Jun 12 14:16:16 hetznix graylogctl[4970]: 2017-06-12 14:16:16,046 INFO : org.graylog2.inputs.InputStateListener - Input [Beats/593d5937bfc060437b083017] is now STARTING
Jun 12 14:16:16 hetznix graylogctl[4970]: 2017-06-12 14:16:16,055 ERROR: org.graylog2.shared.inputs.InputLauncher - The [org.graylog.plugins.beats.BeatsInput] input with ID <593d5937bfc060437b083017> misfired. Reason: Address already in use.
Jun 12 14:16:16 hetznix graylogctl[4970]: org.graylog2.plugin.inputs.MisfireException: org.graylog2.plugin.inputs.MisfireException: org.jboss.netty.channel.ChannelException: Failed to bind to: /0.0.0.0:5044
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.graylog2.plugin.inputs.MessageInput.launch(MessageInput.java:156) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.graylog2.shared.inputs.InputLauncher$1.run(InputLauncher.java:84) [graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:176) [graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_131]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_131]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_131]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_131]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]
Jun 12 14:16:16 hetznix graylogctl[4970]: Caused by: org.graylog2.plugin.inputs.MisfireException: org.jboss.netty.channel.ChannelException: Failed to bind to: /0.0.0.0:5044
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.graylog2.plugin.inputs.transports.NettyTransport.launch(NettyTransport.java:155) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.graylog2.plugin.inputs.MessageInput.launch(MessageInput.java:153) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         ... 7 more
Jun 12 14:16:16 hetznix graylogctl[4970]: Caused by: org.jboss.netty.channel.ChannelException: Failed to bind to: /0.0.0.0:5044
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.jboss.netty.bootstrap.ServerBootstrap.bind(ServerBootstrap.java:272) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.graylog2.plugin.inputs.transports.NettyTransport.launch(NettyTransport.java:141) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.graylog2.plugin.inputs.MessageInput.launch(MessageInput.java:153) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         ... 7 more
Jun 12 14:16:16 hetznix graylogctl[4970]: Caused by: java.net.BindException: Address already in use
Jun 12 14:16:16 hetznix graylogctl[4970]:         at sun.nio.ch.Net.bind0(Native Method) ~[?:1.8.0_131]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at sun.nio.ch.Net.bind(Net.java:433) ~[?:1.8.0_131]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at sun.nio.ch.Net.bind(Net.java:425) ~[?:1.8.0_131]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223) ~[?:1.8.0_131]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74) ~[?:1.8.0_131]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.jboss.netty.channel.socket.nio.NioServerBoss$RegisterTask.run(NioServerBoss.java:193) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.jboss.netty.channel.socket.nio.AbstractNioSelector.processTaskQueue(AbstractNioSelector.java:391) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:315) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.jboss.netty.channel.socket.nio.NioServerBoss.run(NioServerBoss.java:42) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42) ~[graylog.jar:?]
Jun 12 14:16:16 hetznix graylogctl[4970]:         ... 3 more
Jun 12 14:16:16 hetznix graylogctl[4970]: 2017-06-12 14:16:16,063 INFO : org.graylog2.inputs.InputStateListener - Input [Beats/593d5937bfc060437b083017] is now FAILED

```


## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->

* Graylog Version: 2.3.0-alpha.3+c795033
* Elasticsearch Version: 5.4.0
* MongoDB Version: 3.2.9
* Operating System: NixOS 17.09

--- a/graylog2-server/src/main/java/org/graylog2/filters/ExtractorFilter.java
+++ b/graylog2-server/src/main/java/org/graylog2/filters/ExtractorFilter.java
@@ -22,6 +22,9 @@ import com.google.common.eventbus.Subscribe;
 import org.graylog2.database.NotFoundException;
 import org.graylog2.inputs.Input;
 import org.graylog2.inputs.InputService;
+import org.graylog2.inputs.extractors.events.ExtractorCreated;
+import org.graylog2.inputs.extractors.events.ExtractorDeleted;
+import org.graylog2.inputs.extractors.events.ExtractorUpdated;
 import org.graylog2.plugin.Message;
 import org.graylog2.plugin.filters.MessageFilter;
 import org.graylog2.plugin.inputs.Extractor;
@@ -34,6 +37,7 @@ import org.slf4j.LoggerFactory;
 import javax.inject.Inject;
 import javax.inject.Named;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.List;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
@@ -98,7 +102,29 @@ public class ExtractorFilter implements MessageFilter {
     @Subscribe
     @SuppressWarnings("unused")
     public void handleInputUpdate(final InputUpdated event) {
+        LOG.debug("Updating extractors cache for input <{}>", event.id());
         scheduler.submit(() -> loadExtractors(event.id()));
+    }
+
+    @Subscribe
+    @SuppressWarnings("unused")
+    public void handleExtractorCreate(final ExtractorCreated event) {
+        LOG.debug("Load extractors for input <{}>", event.inputId());
+        scheduler.submit(() -> loadExtractors(event.inputId()));
+    }
+
+    @Subscribe
+    @SuppressWarnings("unused")
+    public void handleExtractorDelete(final ExtractorDeleted event) {
+        LOG.debug("Removing extractors for input <{}> from extractors cache", event.inputId());
+        scheduler.submit(() -> loadExtractors(event.inputId()   ));
+    }
+
+    @Subscribe
+    @SuppressWarnings("unused")
+    public void handleExtractorUpdate(final ExtractorUpdated event) {
+        LOG.debug("Updating extractors cache for input <{}>", event.inputId());
+        scheduler.submit(() -> loadExtractors(event.inputId()));
     }
 
     private void loadAllExtractors() {
@@ -114,9 +140,8 @@ public class ExtractorFilter implements MessageFilter {
 
         try {
             final Input input = inputService.find(inputId);
-            final List<Extractor> sortedExtractors = inputService.getExtractors(input)
-                    .stream()
-                    .sorted((e1, e2) -> e1.getOrder().intValue() - e2.getOrder().intValue())
+            final List<Extractor> sortedExtractors = inputService.getExtractors(input).stream()
+                    .sorted(Comparator.comparingLong(Extractor::getOrder))
                     .collect(Collectors.toList());
 
             extractors.put(inputId, ImmutableList.copyOf(sortedExtractors));

--- a/graylog2-server/src/main/java/org/graylog2/inputs/InputService.java
+++ b/graylog2-server/src/main/java/org/graylog2/inputs/InputService.java
@@ -93,6 +93,8 @@ public interface InputService extends PersistedService {
 
     Extractor getExtractor(Input input, String extractorId) throws NotFoundException;
 
+    void updateExtractor(Input input, Extractor extractor) throws ValidationException;
+
     void removeExtractor(Input input, String extractorId);
 
     void removeStaticField(Input input, String key);

--- a/graylog2-server/src/main/java/org/graylog2/inputs/InputServiceImpl.java
+++ b/graylog2-server/src/main/java/org/graylog2/inputs/InputServiceImpl.java
@@ -201,6 +201,13 @@ public class InputServiceImpl extends PersistedServiceImpl implements InputServi
     }
 
     @Override
+    public void updateExtractor(Input input, Extractor extractor) throws ValidationException {
+        removeEmbedded(input, InputImpl.EMBEDDED_EXTRACTORS, extractor.getId());
+        embed(input, InputImpl.EMBEDDED_EXTRACTORS, extractor);
+        publishChange(InputUpdated.create(input.getId()));
+    }
+
+    @Override
     public void addStaticField(Input input, final String key, final String value) throws ValidationException {
         final EmbeddedPersistable obj = new EmbeddedPersistable() {
             @Override

--- a/graylog2-server/src/main/java/org/graylog2/inputs/InputServiceImpl.java
+++ b/graylog2-server/src/main/java/org/graylog2/inputs/InputServiceImpl.java
@@ -34,6 +34,9 @@ import org.graylog2.database.PersistedServiceImpl;
 import org.graylog2.events.ClusterEventBus;
 import org.graylog2.inputs.converters.ConverterFactory;
 import org.graylog2.inputs.extractors.ExtractorFactory;
+import org.graylog2.inputs.extractors.events.ExtractorCreated;
+import org.graylog2.inputs.extractors.events.ExtractorDeleted;
+import org.graylog2.inputs.extractors.events.ExtractorUpdated;
 import org.graylog2.plugin.configuration.Configuration;
 import org.graylog2.plugin.database.EmbeddedPersistable;
 import org.graylog2.plugin.database.Persisted;
@@ -197,14 +200,14 @@ public class InputServiceImpl extends PersistedServiceImpl implements InputServi
     @Override
     public void addExtractor(Input input, Extractor extractor) throws ValidationException {
         embed(input, InputImpl.EMBEDDED_EXTRACTORS, extractor);
-        publishChange(InputUpdated.create(input.getId()));
+        publishChange(ExtractorCreated.create(input.getId(), extractor.getId()));
     }
 
     @Override
     public void updateExtractor(Input input, Extractor extractor) throws ValidationException {
         removeEmbedded(input, InputImpl.EMBEDDED_EXTRACTORS, extractor.getId());
         embed(input, InputImpl.EMBEDDED_EXTRACTORS, extractor);
-        publishChange(InputUpdated.create(input.getId()));
+        publishChange(ExtractorUpdated.create(input.getId(), extractor.getId()));
     }
 
     @Override
@@ -331,7 +334,7 @@ public class InputServiceImpl extends PersistedServiceImpl implements InputServi
     @Override
     public void removeExtractor(Input input, String extractorId) {
         removeEmbedded(input, InputImpl.EMBEDDED_EXTRACTORS, extractorId);
-        publishChange(InputUpdated.create(input.getId()));
+        publishChange(ExtractorDeleted.create(input.getId(), extractorId));
     }
 
     @Override

--- a/graylog2-server/src/main/java/org/graylog2/inputs/extractors/events/ExtractorCreated.java
+++ b/graylog2-server/src/main/java/org/graylog2/inputs/extractors/events/ExtractorCreated.java
@@ -1,0 +1,40 @@
+/**
+ * This file is part of Graylog.
+ *
+ * Graylog is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Graylog is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.inputs.extractors.events;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.auto.value.AutoValue;
+import org.graylog.autovalue.WithBeanGetter;
+
+@AutoValue
+@WithBeanGetter
+@JsonAutoDetect
+public abstract class ExtractorCreated {
+    @JsonProperty("input_id")
+    public abstract String inputId();
+
+    @JsonProperty("extractor_id")
+    public abstract String extractorId();
+
+    @JsonCreator
+    public static ExtractorCreated create(@JsonProperty("input_id") String inputId,
+                                          @JsonProperty("extractor_id") String extractorId) {
+        return new AutoValue_ExtractorCreated(inputId, extractorId);
+    }
+}

--- a/graylog2-server/src/main/java/org/graylog2/inputs/extractors/events/ExtractorDeleted.java
+++ b/graylog2-server/src/main/java/org/graylog2/inputs/extractors/events/ExtractorDeleted.java
@@ -1,0 +1,40 @@
+/**
+ * This file is part of Graylog.
+ *
+ * Graylog is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Graylog is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.inputs.extractors.events;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.auto.value.AutoValue;
+import org.graylog.autovalue.WithBeanGetter;
+
+@AutoValue
+@WithBeanGetter
+@JsonAutoDetect
+public abstract class ExtractorDeleted {
+    @JsonProperty("input_id")
+    public abstract String inputId();
+
+    @JsonProperty("extractor_id")
+    public abstract String extractorId();
+
+    @JsonCreator
+    public static ExtractorDeleted create(@JsonProperty("input_id") String inputId,
+                                          @JsonProperty("extractor_id") String extractorId) {
+        return new AutoValue_ExtractorDeleted(inputId, extractorId);
+    }
+}

--- a/graylog2-server/src/main/java/org/graylog2/inputs/extractors/events/ExtractorUpdated.java
+++ b/graylog2-server/src/main/java/org/graylog2/inputs/extractors/events/ExtractorUpdated.java
@@ -1,0 +1,40 @@
+/**
+ * This file is part of Graylog.
+ *
+ * Graylog is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Graylog is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.inputs.extractors.events;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.auto.value.AutoValue;
+import org.graylog.autovalue.WithBeanGetter;
+
+@AutoValue
+@WithBeanGetter
+@JsonAutoDetect
+public abstract class ExtractorUpdated {
+    @JsonProperty("input_id")
+    public abstract String inputId();
+
+    @JsonProperty("extractor_id")
+    public abstract String extractorId();
+
+    @JsonCreator
+    public static ExtractorUpdated create(@JsonProperty("input_id") String inputId,
+                                          @JsonProperty("extractor_id") String extractorId) {
+        return new AutoValue_ExtractorUpdated(inputId, extractorId);
+    }
+}

--- a/graylog2-server/src/main/java/org/graylog2/rest/resources/system/inputs/ExtractorsResource.java
+++ b/graylog2-server/src/main/java/org/graylog2/rest/resources/system/inputs/ExtractorsResource.java
@@ -169,9 +169,8 @@ public class ExtractorsResource extends RestResource {
         final Extractor originalExtractor = inputService.getExtractor(mongoInput, extractorId);
         final Extractor extractor = buildExtractorFromRequest(cer, originalExtractor.getId());
 
-        inputService.removeExtractor(mongoInput, originalExtractor.getId());
         try {
-            inputService.addExtractor(mongoInput, extractor);
+            inputService.updateExtractor(mongoInput, extractor);
         } catch (ValidationException e) {
             LOG.error("Extractor persist validation failed.", e);
             throw new BadRequestException(e);

